nomad/core_sched.go

```diff
@@ -286,7 +288,13 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	gcEval := true
 	var gcAllocIDs []string
 	for _, alloc := range allocs {
-		if !alloc.TerminalStatus() || alloc.ModifyIndex > thresholdIndex {
+		var reschedulePolicy *structs.ReschedulePolicy
+		tg := job.LookupTaskGroup(alloc.TaskGroup)
+
+		if tg != nil {
+			reschedulePolicy = tg.ReschedulePolicy
+		}
+		if !alloc.GCEligible(reschedulePolicy, time.Now(), thresholdIndex) {
```

Comment:
You may have to apply the same trick as: https://github.com/hashicorp/nomad/blob/9c790214f9b3cb333090c2a3e119453eb634e74b/nomad/core_sched.go#L254-L277 such that a dead/stopped/removed job allows the alloc to disappear.